drivers/comparator/comparator_silabs_acmp.c

```diff
@@ -35,33 +36,76 @@ struct acmp_data {
 	void *user_data;
 };
 
-static int acmp_init(const struct device *dev)
+static int acmp_pm_action(const struct device *dev, enum pm_device_action action)
 {
-	int err;
 	const struct acmp_config *config = dev->config;
+	int err;
 
-	/* Enable ACMP Clock */
-	err = clock_control_on(config->clock_dev, (clock_control_subsys_t)&config->clock_cfg);
-	if (err < 0) {
-		return err;
-	}
-
-	err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
-	if (err < 0 && err != -ENOENT) {
-		LOG_ERR("failed to allocate silabs,analog-bus via pinctrl");
-		return err;
+	switch (action) {
+	case PM_DEVICE_ACTION_TURN_ON:
+	case PM_DEVICE_ACTION_RESUME:
+		/* Enable ACMP Clock */
+		err = clock_control_on(config->clock_dev,
+				       (clock_control_subsys_t)&config->clock_cfg);
+		if (err < 0 && err != -EALREADY) {
+			return err;
+		}
+
+		err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+		if (err < 0 && err != -ENOENT) {
+			LOG_ERR("failed to allocate silabs,analog-bus via pinctrl");
+			return err;
+		}
+
+		if (action == PM_DEVICE_ACTION_TURN_ON) {
+			/* Initialize the ACMP */
+			ACMP_Init(config->base, &config->init);
+
+			/* Configure the ACMP Input Channels */
+			ACMP_ChannelSet(config->base, config->input_negative,
+					config->input_positive);
+
+			/* After the TURN_ON action has executed, the comparator should not be
+			 * enabled. It was temporarily enabled by the init function to perform
+			 * input configuration since the INPUTCTRL register has the SYNC type,
+			 * disable it again here.
+			 */
+			ACMP_Disable(config->base);
+		} else {
+			ACMP_Enable(config->base);
+		}
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		if (IS_ENABLED(CONFIG_PM_DEVICE)) {
+			ACMP_Disable(config->base);
+
+			err = clock_control_off(config->clock_dev,
+						(clock_control_subsys_t)&config->clock_cfg);
+			if (err < 0) {
+				return err;
+			}
+
+			err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_SLEEP);
+			if (err < 0 && err != -ENOENT) {
+				return err;
+			}
+		}
+		break;
+	default:
+		return -ENOTSUP;
 	}
 
-	/* Initialize the ACMP */
-	ACMP_Init(config->base, &config->init);
+	return 0;
+}
 
-	/* Configure the ACMP Input Channels */
-	ACMP_ChannelSet(config->base, config->input_negative, config->input_positive);
+static int acmp_init(const struct device *dev)
+{
+	const struct acmp_config *config = dev->config;
 
 	/* Initialize the irq handler */
 	config->irq_init();
 
-	return 0;
+	return pm_device_driver_init(dev, acmp_pm_action);
 }
```

Comment:
Just an idea, I wonder if we should not define `acmp_init##inst()` rather than `acmp_irq_init##inst()`.

In all the other drivers `config->irq_init()` was called in the middle of init function. But in this one it may make sense.

Reply:
The reason for the `irq_init_##inst()` is that it has to be unique because of IRQ handler linker section quirks, and it is usually a bit smaller than the init function, so it saves ROM to "reuse" the init function and only create "duplicate" functions for `irq_init_##inst()`